### PR TITLE
Ability to add elements to Imager after initialisation

### LINKED ISSUE #110: Imager use with single-page application?
I'm looking into converting a section of a site (http://kvdesign.net/work) into a single-page "application" using Backbone.Marionette. This would (I hope!) make it easier to change it into something along the lines of the Google image-search results page.

The current site uses a slightly modified Imager throughout, with the image variations being created and cached dynamically with a custom ExpressionEngine extension/plugin. This part I can bypass easily enough, but would it be best to... 

1) Render the html as originally output and create the default Imager "onrender" and let it do its thing, since the initial DOM would be loaded before the necessary image elements are there, or

2) Incorporate the Imager's functionality directly into my application code and render the "correct" image from the start, or

3) Bag the whole Backbone idea and live with some confusing jQuery spaghetti code. ;-)

There seem to be advantages and disadvantages to each, with #1 only having the disadvantage of an additional screen redraw. #2 gets rid of that, but then would require the full Imager anyway to take care of screen rotation, etc, unless I re-rendered the "page" with different image srcs. (Plus I'd have to figure it all out.) #3 is probably the easiest to wrap my head around, but lacks elegance.

I'm not worried about older browsers, since her target audience mostly uses up-to-date hardware and software, including retina-capable stuff. Things just have to look good and have a bit of UI "goodness" that designers and their clients like to see. (The entire site was originally conceived as a single-page, parallax site.)

Any suggestions?

### PR DESCRIPTION
**Pull request by WebSeed**
This is very similar to (and inspired by) https://github.com/BBC-News/Imager.js/pull/92, with a slight change in the interface and the addition of a few tests.

closes #92
closes #110
refs #24

### CODE REVIEW

**Comment by WebSeed** [comment]
@oncletom Sorry to be a pain - any update on accepting this PR?

File: Imager.js [hunk 1]
```diff
@@ -106,7 +107,7 @@
         }
 
         this.viewportHeight   = doc.documentElement.clientHeight;
-        this.selector         = opts.selector || '.delayed-image-load';
+        this.selector         = !elements ? (opts.selector || '.delayed-image-load') : null;
```
Comment (oncletom) [inline Imager.js]: :+1:

File: Imager.js [hunk 2]
```diff
@@ -142,30 +143,36 @@
...
-        setTimeout(function(){
+        setTimeout(function () {
             self.init();
         }, 0);
-    }
+    };
+
+    Imager.prototype.add = function (elementsOrSelector) {
+
+        elementsOrSelector = elementsOrSelector || this.selector;
```
Comment (oncletom) [inline Imager.js]: Because of the *real change no1*, `this.selector` can be null. To be honest I would allow an empty call like `imgr.add()`: it would rematch existing responsive images handled by Imager as no deduplication is involved (which is another problem, but I don't want that to be eased up).

What do you think?

File: Imager.js [hunk 3]
```diff
@@ -142,30 +143,36 @@
...
+        var elements = typeof elementsOrSelector === 'string' ?
+            document.querySelectorAll(elementsOrSelector) : // Selector
+            elementsOrSelector; // Elements (NodeList or array of Nodes)
+
+        if (elements.length) {
+            this.divs = this.divs.concat(applyEach(elements, returnFn));
```
Comment (oncletom) [inline Imager.js]: I would apply `changeDivsToEmptyImages` only to the new divs, just to sake of saving some iterations.

File: test/unit/core.js [hunk 4]
```diff
@@ -95,6 +95,58 @@ describe('Imager.js', function () {
...
+            });
+        });
+
+        it('should add additional images based on Node array passed in', function (done) {
+            fixtures = loadFixtures('add');
+            var imgr = new Imager();
+            imgr.ready(function () {
+                var elements = document.querySelectorAll('#test-case div');
+                var a = [];
+                for (var i = 0, n; n = elements[i]; ++i) a.push(n);
```
Comment (oncletom) [inline test/unit/core.js]: Use `Imager.applyEach` static method, it will do the same job and keep the test cleaner.

**Comment by oncletom** [comment]
Sorry for the delay @WebSeed, a bit chasing time at the moment.

I have a made a couple of comments. Let me know what you think about them. So far so good otherwise :-)

**Comment by WebSeed** [comment]
Hi @oncletom, no problem regarding the delay - thanks for the feedback. I think I've addressed the comments.

**Comment by oncletom** [comment]
Thanks @WebSeed I will review your changes tomorrow and publish it :-)

File: Imager.js (hunk 2 above)
Comment (WebSeed) [inline Imager.js]: The problem here is with post-loading elements, i.e. ajax, as this could introduce new elements with a class of ```delayed-image-load``` i.e.

```
var imgr = new Imager(); // default selector is 'delayed-image-load'
...
// ajax some elements into the DOM with class of 'delayed-image-load'
...
imgr.add(); // doesn't pick up elements unless we allow default selector to apply
```

This would be  a workaround for users but doesn't seem right:

```
imgr.add(imgr.selector);
```

Allowing a "no argument" version of ```add``` is the best possible solution I can think of.

Comment (oncletom) [inline Imager.js]: @WebSeed I see what you mean. Here are my two concerns:

- a one line setup is nice and implicit values can be acceptable at that stage – having a mutating function without parameters seems odd
- a call to `.add()` will rematch and add existing responsive images in the pool – if you call it 5 times without adding any new image, the `imgr.divs` length will be multiplied by 5 anyway

If elements are loaded in the DOM, *you* have a reference to those elements, and they should be explicitly passed as an argument of `add` instead of *expecting* them to be matched by the initial selector.

It is a more functional-ish approach, more rigorous but less pragmatic though.

tl;dr: an empty call to `.add()` seems to me to be rather an `.update()` method.

Comment (WebSeed) [inline Imager.js]: > having a mutating function without parameters seems odd.

Yes agree with this.

> a call to .add() will rematch and add existing responsive images in the pool – if you call it 5 times without adding any new image, the imgr.divs length will be multiplied by 5 anyway

Doesn't the matching selector class is dropped from transformed elements i.e.

```
<div class='delayed-image-load' data-src='...'></div>
```

To...
```
<img class='image-replace' src='...'>
```

So calling ```add()``` multiple times won't increase the size of the ```imgr.div``` array.

Comment (oncletom) [inline Imager.js]: Ah yes indeed, I somewhat had in mind the existing classes were forwared to the responsive image… which is not the case.

We can survive with the oddity for the moment then :-)

Thanks!

**Comment by oncletom** [comment]
Merged and shipped in `v0.5.0`. Thanks @WebSeed for the contribution and your patience :-)

**Comment by WebSeed** [comment]
Cheers 😃
